google/cloud/spanner/internal/connection_impl.cc

```diff
@@ -459,22 +452,23 @@ StatusOr<std::unique_ptr<SessionHolder>> ConnectionImpl::GetSession(
     if (!response) {
       return response.status();
     }
-    session_name = std::move(*response->mutable_name());
+    session = google::cloud::internal::make_unique<Session>(
+        std::move(*response->mutable_name()));
   }
 
+  if (dissociate_from_pool) {
+    // Uses the default deleter; the Session is not returned to the pool.
+    return {std::move(session)};
+  }
   // TODO(#409) take a (weak) reference to `this` to avoid use-after-free.
-  auto deleter = dissociate_from_pool
-                     ? std::function<void(std::string)>()
-                     : [this](std::string session) {
-                         this->ReleaseSession(std::move(session));
-                       };
-  return ::google::cloud::internal::make_unique<SessionHolder>(
-      std::move(session_name), std::move(deleter));
+  return SessionHolder(session.release(), [this](Session* session) {
+    this->ReleaseSession(session);
```

Comment:
Does something need to `delete session` somewhere? 

 #IAmNotAUniquePtrDeleterExpert

Reply:
Nah, the `ReleaseSession()` call transfers ownership of the pointer to the `std::vector<std::unique_ptr<Session>>`, which will actually do the delete (eventually).

Reply:
I see now. There are two different unique_ptr's holding `Session`s. One of them uses this custom deleter, the other doesn't. And we transfer ownership between them. Ack.